### PR TITLE
Rework major/minor ticks for log scales

### LINKED ISSUE #3303: Setting custom ticks in log plots with minor ticks enabled has confusing behavior
Say, for example, that I want to make a simple log-log plot.

```
x = range(0.1, 100, length=1000);
y = x.^2
```
And then plot it on a log log scale
```
p = plot(x, y, axis=:log, minorticks=10, minorgrid=true)
```
This will work, but produces a plot whose major ticks may not be where I want them (I.E not whole numbers, or whatever)

So if I go and add in x/y tick marks to correspond to the major locations, I.E
```
p = plot( ... , xticks=[5e-1, 5e0, 5e1, 5e2])
```
This works fine. But let's say now I want to just plot a sub-range of that data.
```
p = plot(x, y, axis=:log, minorticks=10, minorgrid=true, xticks=[5e-1, 5e0, 5e1, 5e2])
xlims!(1,10)
```
This will produce the following error:
```
julia> PlotTest.test_plot()
ERROR: MethodError: no method matching iterate(::Nothing)
Closest candidates are:
  iterate(::ExponentialBackOff) at error.jl:252
  iterate(::ExponentialBackOff, ::Any) at error.jl:252
  iterate(::Base.AsyncGenerator, ::Base.AsyncGeneratorState) at asyncmap.jl:382
  ...
Stacktrace:
 [1] axis_drawing_info(::Plots.Subplot{Plots.GRBackend}, ::Symbol) at C:\Users\fergu\.julia\packages\Plots\5ItHH\src\axes.jl:668
 [2] gr_draw_axis(::Plots.Subplot{Plots.GRBackend}, ::Symbol, ::Array{Float64,1}) at C:\Users\fergu\.julia\packages\Plots\5ItHH\src\backends\gr.jl:1222
 [3] gr_draw_axes(::Plots.Subplot{Plots.GRBackend}, ::Array{Float64,1}) at C:\Users\fergu\.julia\packages\Plots\5ItHH\src\backends\gr.jl:1216
 [4] gr_display(::Plots.Subplot{Plots.GRBackend}, ::Measures.Length{:mm,Float64}, ::Measures.Length{:mm,Float64}, ::Array{Float64,1}) at C:\Users\fergu\.julia\packages\Plots\5ItHH\src\backends\gr.jl:881
 [5] gr_display(::Plots.Plot{Plots.GRBackend}, ::String) at C:\Users\fergu\.julia\packages\Plots\5ItHH\src\backends\gr.jl:596
 [6] _show(::Base.GenericIOBuffer{Array{UInt8,1}}, ::MIME{Symbol("image/svg+xml")}, ::Plots.Plot{Plots.GRBackend}) at C:\Users\fergu\.julia\packages\Plots\5ItHH\src\backends\gr.jl:1777
 [7] display(::VSCodeServer.InlineDisplay, ::MIME{Symbol("image/svg+xml")}, ::Plots.Plot{Plots.GRBackend}) at c:\Users\fergu\.vscode\extensions\julialang.language-julia-1.1.17\scripts\packages\VSCodeServer\src\display.jl:0   
 [8] display(::VSCodeServer.InlineDisplay, ::String, ::Any) at .\multimedia.jl:216
 [9] display(::VSCodeServer.InlineDisplay, ::Plots.Plot{Plots.GRBackend}) at c:\Users\fergu\.vscode\extensions\julialang.language-julia-1.1.17\scripts\packages\VSCodeServer\src\display.jl:108
 [10] display(::Any) at .\multimedia.jl:328
 [11] test_plot() at K:\data\ShockDt\PlotTest.jl:8
 [12] top-level scope at REPL[102]:1
```

Likewise, if I comment out the `xlims!` line and instead do something else like...

```
# Note that the ticks have been changed to span the range of x that I want to restrict to via xlims!
p = plot(x, y, axis=:log, minorticks=10, minorgrid=true, xticks=[1e-1, 5e0, 1e1])
```
This works fine, but if I reintroduce that `xlims!` call I get the same error, even though the the ticks span the whole range I want to view. (And, in fact, it seems that the ticks don't even have to span the range of the data)

Additionally, if I remove the `minorticks` and `minorgrid` arguments, this works fine. It also works as expected if I keep the minor axis arguments but remove the axis scale specification.

I would expect that setting ticks and lims to not have a dependence on each other, or on the scale type or whether a minor grid exists. I.E - if a tick falls outside of the range specified, it would just not be plotted. If the limits are greater than the specified ticks there are either no ticks plotted (which seems to already be the case), or an error indicating that the ticks do not span the plotted range to be emitted.

### PR DESCRIPTION
**Pull request by t-bltg**
This PR is a proposal for generating automatic ticks on `log` based scale family (`log`, `log2`, `log10`).
Currently, exponents are not integer based.
Furthermore, minor ticks are generated using arbitrary (`5`) intervals for both linear and log scales.

This PR:
- adds automatic computation of major ticks based on integer exponents for log scales (decades for `log10`);
- computes sub minor ticks when transitioning from non unit order of magnitude,  e.g.  `10¹ -> 10³` instead of `10¹ -> 10² -> 10³` on major ticks for a `log10` based scale (see below for an example).

Example:
```julia
using Plots; gr()
x = y = Float64[1]
for i ∈ range(-5, 6, step=1)
  append!(x, 2 * 10.0^i:10.0^i:10.0^(i + 1) |> collect)
end
plot(
  x, y, axis=:log10, minorgrid=true, legend=false,
  minorticks=true, shape=:cross, ms=2, lw=.5, dpi=300
)
png("loglog")
```
Without PR (open the link for high res minorticks grid):
loglog-cmp-NOK

With PR (open the link for high res minorticks grid):
loglog-cmp-OK

Closes https://github.com/JuliaPlots/Plots.jl/issues/3318
Fixes #3303

### CODE REVIEW

File: src/axes.jl [hunk 1]
```diff
@@ -169,12 +185,17 @@ function optimal_ticks_and_labels(ticks, alims, scale, formatter)
 
     # get a list of well-laid-out ticks
     if ticks === nothing
-        scaled_ticks = optimize_ticks(
-            sf(amin),
-            sf(amax);
-            k_min = 4, # minimum number of ticks
-            k_max = 8, # maximum number of ticks
-        )[1]
+        if scale in (:ln, :log2, :log10)
```
Comment (BeastyBlacksmith) [inline src/axes.jl:188]:
```suggestion
        if scale in _logScales
```

File: src/axes.jl [hunk 2]
```diff
@@ -169,12 +185,17 @@ function optimal_ticks_and_labels(ticks, alims, scale, formatter)
...
+            base = scale == :ln ? ℯ : (scale == :log2 ? 2. : 10.)
```
Comment (BeastyBlacksmith) [inline src/axes.jl:189]:
```suggestion
            base = _logScaleBases[scale]
```

File: src/axes.jl [hunk 3]
```diff
@@ -348,13 +370,33 @@ function get_minor_ticks(sp, axis, ticks)
     last_step = ticks[end] - ticks[end-1]
     ticks =  [ticks[1] - first_step/ratio; ticks; ticks[end] + last_step*ratio]
 
-    #Default to 5 intervals between major ticks
-    n = typeof(axis[:minorticks]) <: Integer && axis[:minorticks] > 1 ? axis[:minorticks] : 5
+    # Default to 5 intervals between major ticks on a linear scale,
+    # base - 1 for integral based log scale and 0 for ℯ based log scale
+    integral_log = scale in (:log2, :log10)
+    base = scale == :log2 ? 2 : 10
```
Comment (BeastyBlacksmith) [inline src/axes.jl:376]:
```suggestion
    base = _logScaleBases[scale]
```

**Comment by t-bltg** [comment]
@BeastyBlacksmith thanks for the review, I will apply the changes right away.

In the mean time, the CI build failed (e.g. https://github.com/JuliaPlots/PlotReferenceImages.jl/blob/master/Plots/gr/1.11.2/ref5.png), can you please suggest a procedure to regenerate the reference images ?

**Comment by BeastyBlacksmith** [comment]
You need to create the image locally and then provide a PR against PlotReferenceImages.jl, like https://github.com/JuliaPlots/PlotReferenceImages.jl/pull/81

**Comment by t-bltg** [comment]
> You need to create the image locally and then provide a PR against PlotReferenceImages.jl, like [JuliaPlots/PlotReferenceImages.jl#81](https://github.com/JuliaPlots/PlotReferenceImages.jl/pull/81)

Thanks, will try.
> How do the ticks look like if I only plot in a narrow range, lets say two order of magnitudes?

Here is a narrower example:
loglog-cmp

**Review by BeastyBlacksmith — APPROVED** [review]
I think this looks good, but would be good to have some other eyes on this since ticks are an opinionated topic

File: src/axes.jl [hunk 4]
```diff
@@ -732,7 +774,10 @@ function axis_drawing_info(sp, letter)
                 end
             end
 
-            if !(ax[:minorticks] in (:none, nothing, false)) || ax[:minorgrid]
+            if (
+                (!(ax[:minorticks] in (:none, nothing, false)) || ax[:minorgrid]) &&
+                !isnothing(minor_ticks)
```
Comment (t-bltg) [inline src/axes.jl:779]: Added a guard against minor_ticks === nothing.

Comment (BeastyBlacksmith) [reply src/axes.jl:779]:
```suggestion
                minor_ticks !== nothing
```

**Comment by t-bltg** [comment]
> I think this looks good, but would be good to have some other eyes on this since ticks are an opinionated topic

Yes, better have other reviews (not bug 100% free e.g. when  `get_minor_ticks` returns nothing https://github.com/JuliaPlots/Plots.jl/blob/master/src/axes.jl#L345).

File: src/axes.jl (hunk 4 above)
Comment (t-bltg) [reply src/axes.jl:779]: Is there a reason for not using the builtin `isnothing` ?

**Comment by BeastyBlacksmith** [comment]
Would be also good if you can check what is the influence on https://github.com/JuliaPlots/Plots.jl/issues/3303 of this PR

File: src/axes.jl (hunk 4 above)
Comment (BeastyBlacksmith) [reply src/axes.jl:779]: the compiler loves `===` checks

**Comment by t-bltg** [comment]
> Would be also good if you can check what is the influence on #3303 of this PR

```julia
using Plots; gr()

x = range(0.1, 100, length=1000)
y = x.^2

p = plot(x, y, axis=:log, minorticks=10, minorgrid=true, xticks=[5e-1, 5e0, 5e1, 5e2])
xlims!(1,10)
png("test0.png")

p = plot(x, y, axis=:log, minorticks=10, minorgrid=true, xticks=[1e-1, 5e0, 1e1])
png("test1.png")
```

No errors, ticks correct.

File: src/axes.jl (hunk 4 above)
Comment (t-bltg) [reply src/axes.jl]: Updated.

**Comment by t-bltg** [comment]
It totally missed https://github.com/JuliaPlots/Plots.jl/issues/3318 when I wrote this PR. Maybe we can merge the best of both worlds and create something even better ?

**Comment by BeastyBlacksmith** [comment]
Sure, if you find inspiration there to improve this PR. Just indicate when you think it is done.

**Comment by BeastyBlacksmith** [comment]
Another one you might check is https://github.com/JuliaPlots/Plots.jl/issues/3502

**Comment by t-bltg** [comment]
> Another one you might check is #3502

I think that nothing can be done here since gr does not support `:ln` scale:
```julia
┌ Warning: scale ln is unsupported with Plots.GRBackend().  Choose from: [:identity, :log10]
└ @ Plots [...]/src/args.jl:1260
```
Also, there is no MWE to reproduce (and the screenshot is too small, unreadable).


> Sure, if you find inspiration there to improve this PR. Just indicate when you think it is done.

I've looked into https://github.com/JuliaPlots/Plots.jl/issues/3318, I think nothing can be added here, so I think I'm done.

**Comment by BeastyBlacksmith** [comment]
@daschw, @yha could you have a look at this?

**Comment by isentropic** [comment]
I think the changes should go to Plot utils under `optimize_ticks` function instead. Perhaps overload the function or create `optimize_ticks_log`

**Comment by isentropic** [comment]
Getting better tick locations is the job of the `optimize_ticks` function. Let us keep it like that instead of expanding the Plots.jl codebase more. Delegation to PlotUtils would lead to easier testing too.

**Comment by isentropic** [comment]
Better yet, is to find optimum weights and params to the [optimize_ticks](function optimize_ticks(https://github.com/JuliaPlots/PlotUtils.jl/blob/4007370cd3c3b9a6cdc62871687dac21364190ff/src/ticks.jl#L130) function so that the returned ticks prefer "ten divisible ticks". Current solution feels like a hack. I believe by using the optimize_ticks (which is smart, perhaps too much even) function properly we can achieve the same result.

**Comment by daschw** [comment]
Looks really great, but I agree with @isentropic that the code for finding tick locations should be in PlotUtils.jl.

**Comment by t-bltg** [comment]
Hi all,
Sorry for the late reply.
I'm OK with rewriting stuff to go into https://github.com/JuliaPlots/PlotUtils.jl by the end of the week.

However I think we should bypass `optimize_ticks`'s logic rather that try to bury every functionality into one method.

The essence of this PR is to have `9` intervals per decade between each major tick for a `log10` scale e.g. for two major ticks at `10^0`  and `10^1` we should get in between `2, 3, 4, 5, 6, 7, 8, 9`. For other scales (`log2` and `log` the choice is arbitrary and can be left to the default 5 intervals.

**Comment by t-bltg** [comment]
Superseded by https://github.com/JuliaPlots/Plots.jl/pull/3577, closing.

**Pull request closed without merging**